### ISSUE TITLE
[suggestion] Python GDAL bindings

### ISSUE THREAD

**Issue by Daniel-M** [OSGeo/gdal#822]
Hi GDAL developers.
I appreciate your work in developing this awesome library/tool. The past days I've been working for hours to get compiled GDAL python bindings running on virtual environments and docker containers. I always encountered an error related to `_gdal_array` module not to be found when deploying the app build upon gdal. After an intensive search it turns out that the package `numpy` and python headers are a prerequisite to get GDAL fully functional. I found this at the link [https://pypi.org/project/GDAL/](https://pypi.org/project/GDAL/) which isn't related at all with my problem in search engines. As far as I know, this gotcha is not documented at the official [www.gdal.org](www.gdal.org) web page, nor described in the Readme of this repo or the official Wiki. I just came by to say that if you are responsible for the deployment of the GDAL module to PyPI, add the `numpy` module as a requirement , or at least make it show message to stdout which warns about the missing `_gdal_array` error related to a lack of numpy or the python-dev headers.

On the other hand, the `./configure --with-python=...` does not check if  numpy is available in the system or for the indicated python binary.

The error raises when trying to access the raster bands of a gdal.Dataset, and might look like this,

```
  File "/usr/local/lib/python3.5/dist-packages/osgeo/gdal.py", line 2549, in WriteArray
    from . import gdalnumeric
  File "/usr/local/lib/python3.5/dist-packages/osgeo/gdalnumeric.py", line 1, in <module>
    from osgeo.gdal_array import *
  File "/usr/local/lib/python3.5/dist-packages/osgeo/gdal_array.py", line 28, in <module>
    _gdal_array = swig_import_helper()
  File "/usr/local/lib/python3.5/dist-packages/osgeo/gdal_array.py", line 20, in swig_import_helper
    import _gdal_array
ImportError: No module named '_gdal_array
```

As a comparison, here's how `/usr/local/lib/python3.5/dist-packages/osgeo/` looks when 
numpy is not availabe 
```
root@bc5c55cd2daf:/# ls -la /usr/local/lib/python3.5/dist-packages/osgeo/
total 6068
drwxr-sr-x 3 root staff    4096 Jul 31 17:34 .
drwxrwsr-x 1 root staff    4096 Aug  1 00:52 ..
-rw-r--r-- 1 root staff     777 Apr 14  2017 __init__.py
drwxr-sr-x 2 root staff    4096 Jul 31 17:34 __pycache__
-rwxr-xr-x 1 root staff 2049776 Jul 31 17:33 _gdal.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  217864 Jul 31 17:33 _gdalconst.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  323816 Jul 31 17:33 _gnm.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff 2165920 Jul 31 17:33 _ogr.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  932960 Jul 31 17:33 _osr.cpython-35m-x86_64-linux-gnu.so
-rw-r--r-- 1 root staff  139138 Jul 31 17:33 gdal.py
-rw-r--r-- 1 root staff   18841 Jul 31 17:34 gdal_array.py
-rw-r--r-- 1 root staff   15338 Aug 16  2017 gdalconst.py
-rw-r--r-- 1 root staff      51 Apr 14  2017 gdalnumeric.py
-rw-r--r-- 1 root staff   10868 Jul 31 17:34 gnm.py
-rw-r--r-- 1 root staff  252079 Jul 31 17:33 ogr.py
-rw-r--r-- 1 root staff   46183 Aug 16  2017 osr.py
```
Here's how it looks when `numpy` is available prior GDAL installation,

```
root@e3fbb273422d:/app# ls -la /usr/local/lib/python3/dist-packages/osgeo/
total 6348
drwxr-sr-x 3 root staff    4096 Aug  1 18:53 .
drwxr-sr-x 5 root staff    4096 Aug  1 18:53 ..
-rw-r--r-- 1 root staff     777 Sep 15  2017 __init__.py
drwxr-sr-x 2 root staff    4096 Aug  1 18:53 __pycache__
-rwxr-xr-x 1 root staff 2049808 Aug  1 18:53 _gdal.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  286560 Aug  1 18:53 _gdal_array.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  217880 Aug  1 18:53 _gdalconst.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  323840 Aug  1 18:53 _gnm.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff 2165936 Aug  1 18:53 _ogr.cpython-35m-x86_64-linux-gnu.so
-rwxr-xr-x 1 root staff  932960 Aug  1 18:53 _osr.cpython-35m-x86_64-linux-gnu.so
-rw-r--r-- 1 root staff  139138 Aug  1 18:52 gdal.py
-rw-r--r-- 1 root staff   18867 Aug  1 18:52 gdal_array.py
-rw-r--r-- 1 root staff   15338 Sep 15  2017 gdalconst.py
-rw-r--r-- 1 root staff      51 Sep 15  2017 gdalnumeric.py
-rw-r--r-- 1 root staff   10868 Aug  1 18:52 gnm.py
-rw-r--r-- 1 root staff  252079 Aug  1 18:52 ogr.py
-rw-r--r-- 1 root staff   46183 Sep 15  2017 osr.py
```
## Operating system

Debian testing with linux kernel 4.16

## GDAL version and provenance

Source for GDAL 2.2.2

**Comment by rouault** [comment]
Where would you this to be documented ? This is documented in https://github.com/OSGeo/gdal/blob/master/gdal/swig/python/README.txt

**Comment by Daniel-M** [comment]
Thanks for your fast reply. I can't tell you where to put the documentation, But I'm pointing out that it is not straight forward to find it, to the point that is the first time I've seen such README under the GDAL source tree, is not a straigthforward place to look for it, at least for me.

**Comment by rouault** [comment]
Well, if you can't tell where you would want to see it mentionned, it is going to be difficult for us to know how to improve ;-)

What if this would be displayed when building the bindings ?
```
$ python setup.py build
WARNING: numpy not available!  Array support will not be enabled
running build
running build_py
running build_ext
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I../../port -I../../gcore -I../../alg -I../../ogr/ -I../../ogr/ogrsf_frmts -I../../gnm -I../../apps -I/usr/include/python2.7 -I. -I/home/even/gdal/git/gdal/gdal/include -c gdal_python_cxx11_test.cpp -o gdal_python_cxx11_test.o
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I../../port -I../../gcore -I../../alg -I../../ogr/ -I../../ogr/ogrsf_frmts -I../../gnm -I../../apps -I/usr/include/python2.7 -I. -I/home/even/gdal/git/gdal/gdal/include -c gdal_python_cxx11_test.cpp -o gdal_python_cxx11_test.o -std=c++11
running build_scripts
```

**Comment by Daniel-M** [comment]
Most Python users would try to install GDAL from PyPI with a command like `pip install GDAL`. Doing so doesn't install `numpy` as a requirement. If an hypothetical user wants to use gdal_array capabilities and does install GDAL vía PyPI, might encounter with a bitter surprise as I did. That's why I wrote `if you are responsible for the deployment of the GDAL module to PyPI,` in my initial issue comment. Maybe adding `numpy` as a requirement at PyPI platform or putting explicit section on `Installing GDAL bindings` with language specific instructions would make things a little bit easier for some users, in particular, non experienced users.
